pkg/diff/sql_generator.go

```diff
@@ -1332,6 +1422,106 @@ func (a *attachPartitionSQLVertexGenerator) GetDeleteDependencies(_ schema.Table
+	// This is used to identify if the owning table is being dropped (meaning
+	// the foreign key can be implicitly dropped)
+	deletedTablesByName map[string]schema.Table
+	// addedTablesByName is a map of table name to the added tables
+	// This is used to identify if hazards are necessary
+	addedTablesByName map[string]schema.Table
+	// indexInOldSchemaByTableName is a map of index name to the index in the old schema
+	// This is used to force the foreign key constraint to be dropped before the index it depends on is dropped
+	indexInOldSchemaByTableName map[string][]schema.Index
+	// indexesInNewSchemaByTableName is a map of index name to the index
+	// Same as above but for adds and after
+	indexesInNewSchemaByTableName map[string][]schema.Index
+}
+
+func (f *foreignKeyConstraintSQLVertexGenerator) Add(con schema.ForeignKeyConstraint) ([]Statement, error) {
+	var hazards []MigrationHazard
+	_, isOnNewTable := f.addedTablesByName[con.OwningTableUnescapedName]
+	_, isReferencedTableNew := f.addedTablesByName[con.ForeignTableUnescapedName]
+	if con.IsValid && (!isOnNewTable || !isReferencedTableNew) {
+		hazards = append(hazards, MigrationHazard{
+			Type: MigrationHazardTypeAcquiresShareRowExclusiveLock,
+			Message: "This will lock writes to the owning table and referenced table while the constraint is being added. " +
+				"Instead, consider adding the constraint as NOT VALID and validating it later.",
+		})
+	}
+	return []Statement{{
+		DDL:     fmt.Sprintf("%s %s", addConstraintPrefix(con.OwningTable, con.EscapedName), con.ConstraintDef),
+		Timeout: statementTimeoutDefault,
+		Hazards: hazards,
+	}}, nil
+}
+
+func (f *foreignKeyConstraintSQLVertexGenerator) Delete(con schema.ForeignKeyConstraint) ([]Statement, error) {
+	// Always generate a drop statement even if the owning table is being deleted. This simplifies the logic a bit because
+	// if the owning table has a circular FK dependency with another table being dropped, we will need to explicitly drop
+	// one of the FK's first
+	return []Statement{{
+		DDL:     dropConstraintDDL(con.OwningTable, con.EscapedName),
+		Timeout: statementTimeoutDefault,
+	}}, nil
+}
+
+func (f *foreignKeyConstraintSQLVertexGenerator) Alter(diff foreignKeyConstraintDiff) ([]Statement, error) {
+	var stmts []Statement
+	if !diff.old.IsValid && diff.new.IsValid {
+		diff.old.IsValid = diff.new.IsValid
+		// We're not keeping track of the other FK attributes, so it's easiest to ensure all other properties are equal
+		// by just modifying the old constraint diff to exclude "NOT VALID", which should make the diffs equal if no
+		// other properties have changed
+		if !strings.HasSuffix(diff.old.ConstraintDef, " NOT VALID") {
+			return nil, fmt.Errorf("expected the old constraint def to be suffixed with NOT VALID: %q", diff.old.ConstraintDef)
+		}
+		diff.old.ConstraintDef = strings.TrimSuffix(diff.old.ConstraintDef, " NOT VALID")
```

Comment:
nit: Is this just so the cmp.Equal on 1482 works? Might make sense to just do that inline on 1482

Reply:
This is a really good question. This is a pattern of the `Alter` functions that I'm switching too. 

The Alter functions need to keep track of what they can successfully alter and what they can't. If they can't successfully resolve a diff, then ideally they return an error to the user. 

I found the easiest way to do this is:
For each alterable difference,
- generate a statement to alter the difference
- update a copy of old to include the difference
Finally, compare old to new. If they are equal, then there were no unresolvable diffs.

With the case of invalid to valid, we need both update the old to be valid and the constraint def in order for the equality check to pass. The `strings.HasSuffix` is more of a confidence check that the constraint def follows the expected format.